### PR TITLE
added intra and inter cosine comparisons (for recommendatins and blen…

### PR DESCRIPTION
**Pull request by kshitijnair**
…d matrices)

### CODE REVIEW

File: src/getCosineSimilarities.ts [hunk 1]
```diff
@@ -19,5 +21,29 @@ export function getCosineSimilarities(listOfRecommendationsWithAttributes: numbe
   }
   const averageCosineSimilarity = cosineSimilaritiesOfAllReccomendations / 45;
```
Comment (kevintab95) [inline src/getCosineSimilarities.ts:22]: Please double check the denominator here

File: src/getCosineSimilarities.ts [hunk 2]
```diff
@@ -19,5 +21,29 @@ export function getCosineSimilarities(listOfRecommendationsWithAttributes: numbe
...
+  for (let i = 0; i < sizeA; i++) {
+    for (let j = i + 1; j < sizeA; j++) {
+      if (j == sizeA) continue;
+      cosineSimilaritiesOfAllReccomendations += similarity(
+        matrixA[i],
+        matrixB[j]
+      );
+    }
+  }
+  const averageCosineSimilarity = cosineSimilaritiesOfAllReccomendations / 45;
```
Comment (kevintab95) [inline src/getCosineSimilarities.ts:46]: Ditto: denominator

File: src/app.ts [hunk 3]
```diff
@@ -12,12 +15,26 @@ const port = 3000;
...
+  const recomMatrix = await getRecomFeatures(spotify, recommendationData);
+
+  // console.log("Recomm Matrix is: ");
+  // console.log(recomMatrix);
+
+  console.log(`\nIntra-list similarity is: ${getCosineSimilarities(recomMatrix).toFixed(5)}`);
+
+  // Test to compare with Spotify Blend generated list
+  const blendMatrix = await fetchPlaylists(spotify, BLEND_PLAYLIST_ID);
+  console.log(`\nInter-List similarity (Spotify-Generated Blend Playlist) is: ${getInterCosineSimilarities(recomMatrix, blendMatrix).toFixed(5)}`);
```
Comment (kevintab95) [inline src/app.ts:31]: Not sure if this is useful, we are comparing the similarity of two different recommendations here (whose seeds were completely different.)

Comment (kevintab95) [reply src/app.ts:31]: Perhaps this is a more useful metric?
https://towardsdatascience.com/evaluation-metrics-for-recommender-systems-df56c6611093#:~:text=three%20recommender%20systems.-,Personalization,recommendations.%20An%20example%20will%20best%20illustrate%20how%20personalization%20is%20calculated.,-Example%20list%20of

**Review by kevintab95 — CHANGES_REQUESTED** [review]
Thanks @kshitijnair! Left a couple of comments PTAL.

File: .env [hunk 4]
```diff
@@ -1 +1 @@
-SPOTIFY_KEY=""
+SPOTIFY_KEY="BQDfd4v8ATYWyJonXgFaA1uq2XsfHmvkNWy1xROYHUdZ-B_49o9fBOam-t3258TuNpMJqnqWbhil0iP9MhtOgeIfcQ-MgEN6f5pXAPf2vRmo4qV0S2gBivZAkVQhyJd6uFBQBKJVuCaMDN-jwzPYpCKiTqCAeUjMOfCnMlg"
```
Comment (viral98) [inline .env:1]: Reverrttt

File: src/getCosineSimilarities.ts [hunk 5]
```diff
@@ -19,5 +21,29 @@ export function getCosineSimilarities(listOfRecommendationsWithAttributes: numbe
...
+/**
+ * Testing function to determine the accuracy of the recommendation engine.
+ * @param matrixA a 2-D array of the matrix for which cosine similarity is calcualted.
+ * @param matrixB a 2-D array of the matrix with which cosine similarity is calcualted.
+ * @returns The cosine similarity of the songs recommended by CF
+ */
+export function getInterCosineSimilarities(matrixA: number[][], matrixB: number[][]) {
+  let cosineSimilaritiesOfAllReccomendations = 0;
+  let sizeA = matrixA.length,
+      sizeB = matrixB.length;
```
Comment (viral98) [inline src/getCosineSimilarities.ts:36]: What is sizeB used for?

File: src/getCosineSimilarities.ts (hunk 2 above)
Comment (viral98) [reply src/getCosineSimilarities.ts:45]: Count the number of elements you are adding and divide by that valuye

**Comment by viral98** [comment]
@kevintab95 @kshitijnair I am merging this one to master, @kshitijnair could you send a follow-up to fix the hardcoded 45 issue? Thanks lads :D

**Pull request merged**